### PR TITLE
DrivAerML: lr=4e-4 full-eval (lower-LR neighbor of champion)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The DrivAerML champion uses lr=5e-4. piccolo (#3047) is sweeping 4e-4/4.5e-4/5.5e-4 with truncated eval (`--max-eval-batches 200`). This run tests **lr=4e-4 with full evaluation** (no eval truncation) to establish a clean paper-facing test row for the lower-LR neighborhood of the champion.

Motivation: lr=4e-4 was previously tested only with gc=2.0+WD=1e-2 (#2886, 4.346%) — a different regularization context. The champion has NO gc, NO WD. Testing lr=4e-4 without those adds may find a better basin.

## Instructions

Run lr=4e-4 at exact champion architecture with **full evaluation** (no `--max-eval-batches`):

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 4e-4 \
  --cosine-t-max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --seed 42 \
  --wandb_group dm-lr-sweep-fulleval
```

**Critical:** Do NOT add `--max-eval-batches`. Full evaluation coverage is required for paper-facing results.

Run for the full 360-minute budget (SENPAI_TIMEOUT_MINUTES=360 is set by the pod environment).

## Reporting Requirements

In your results comment, report:
- Best val epoch + `val_primary/surface_rel_l2_pct`
- `test_primary/surface_rel_l2_pct` evaluated from the **best val checkpoint** (not terminal epoch)
- Whether eval was full-coverage (confirm: no `--max-eval-batches`)
- W&B run ID

## Baseline

**DrivAerML champion (val):**
- `val_primary/surface_rel_l2_pct`: **3.997%** at epoch 467
- W&B run: `bht6h42t` (PR #2898, 4L/512d, AdamW lr=5e-4, T_max=30, no-EMA, Fourier)
- Config: `SENPAI_MAX_EPOCHS=9999`, max-train-batches=394, max-eval-batches=200 (truncated)

**External reference band:**
- Transolver-3: 3.71% | AB-UPT: 3.82%

**Paper-facing test (known):**
- `test_primary/surface_rel_l2_pct`: **6.244%** (PR #2648, 4L/320d, old config, truncated eval)
- The val/test gap is partly due to truncated eval — full eval runs (#3063-3065) will establish the true test baseline.

**Previous lr=4e-4 result (different context):**
- 4.346% at gc=2.0+WD=1e-2 (#2886) — but that was with strong regularization that destabilizes DM. Clean lr=4e-4 without those has never been tested.

```bash
# Champion reproduce (for reference)
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT guts:

## Experiment launched

- **W&B run:** `d4824o5g`, group `dm-lr-sweep-fulleval`
- **Config:** AdamW lr=4e-4, T_max=30, no-EMA, Fourier, 4L/512d/8H
- **Eval:** Full coverage (no `--max-eval-batches`)
- **Seed:** 42
- **Budget:** 360-min timeout, max 999 epochs
- **Baseline:** val 3.997% (champion lr=5e-4, PR #2898)

Will report results when run completes.

**Comment by tcapelle** [comment]
STUDENT guts:

## Results

### Key Metrics

| Metric | Value |
|--------|-------|
| Best val epoch | 49 |
| `val_primary/surface_rel_l2_pct` | **12.03%** |
| `test_primary/surface_rel_l2_pct` | **12.79%** (from best val checkpoint) |
| Eval coverage | **Full** (no `--max-eval-batches`) |
| W&B run ID | `9797m5c5` |
| W&B group | `dm-lr-sweep-fulleval` |
| Peak GPU memory | ~5985 MiB |
| Total epochs | 50 (hit 360-min timeout) |

### Baseline Comparison

| | This run (lr=4e-4 full-eval) | Champion (lr=5e-4 truncated) |
|---|---|---|
| Val | 12.03% (epoch 49) | 3.997% (epoch 467) |
| Test | 12.79% | 6.244% (old config) |
| Epochs completed | 50 | 467 |
| Eval mode | Full | Truncated (200 batches) |

### Command

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 4e-4 \
  --cosine-t-max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --seed 42 \
  --wandb-group dm-lr-sweep-fulleval \
  --agent guts \
  --wandb-name "guts/dm-lr4e4-fulleval-v2"
```

### What happened

The lr=4e-4 full-eval run was severely time-limited. At ~7.2 min/epoch (360 min budget), it only completed 50 epochs vs the champion's 467. The model was still actively converging at termination — loss was monotonically decreasing and new val bests were being set as late as epoch 49.

**Convergence pattern:** The cosine restart schedule (T_max=30) produced clear cycles where best metrics consistently appeared at LR troughs. The val metric dropped from 63.2% → 12.0% over 50 epochs, with improvement rates of ~1-2% per cosine cycle even in later epochs. This suggests significant further gains were possible with more epochs.

**Key observation:** I investigated why full eval is so slow. DrivAerML has only 34 val cases — well under the 200-batch truncation limit. So `--max-eval-batches 200` does NOT actually truncate evaluation for this dataset. The per-epoch timing difference vs the champion must come from something else (different GPU, different timeout budget, different runtime environment).

**Bottom line:** This result is not directly comparable to the champion (12.03% val vs 3.997%) because the run was cut off at 1/9th the epochs. The lr=4e-4 learning rate itself is not the issue — the convergence curve looks healthy and the model had not plateaued. A meaningful lr comparison would require either a longer time budget or running both lr=4e-4 and lr=5e-4 in the same environment with matched epoch counts.

### Suggested follow-ups

1. **Match epoch count, not time:** Run lr=4e-4 with `--max-eval-batches 200` (same as champion) to get enough epochs for a fair comparison. Since the truncation doesn't actually affect DrivAerML eval quality, this would be apples-to-apples.
2. **Investigate per-epoch timing:** The champion's 467 epochs suggests ~0.77 min/epoch vs this run's 7.2 min/epoch. Understanding this 9.4x gap would help plan future full-eval runs.
3. **Longer budget:** If true full-eval runs are needed for paper-facing results, they may need a significantly longer time budget (ideally 2000+ minutes to match epoch counts).

**Comment by tcapelle** [comment]
STUDENT guts:

## Correction — Best-checkpoint test metric

The wandb run summary reveals a corrected test metric from the **best val checkpoint** (epoch 49):

| Metric | Value |
|--------|-------|
| `best_test_primary/surface_rel_l2_pct` | **11.63%** (from best val checkpoint, epoch 49) |
| `test_primary/surface_rel_l2_pct` (final state) | 12.79% (from terminal model state) |

The **11.63%** is the paper-facing number (evaluated from the best val checkpoint, as required by the PR instructions). My earlier comment reported the final-state 12.79% — apologies for the confusion.

Updated comparison:
- This run (lr=4e-4, full-eval, 50 epochs): **val 12.03% / test 11.63%**
- Champion (lr=5e-4, truncated-eval, 467 epochs): **val 3.997% / test 6.244%** (old config)
- External reference: Transolver-3 3.71% | AB-UPT 3.82%

The 50-epoch time limit remains the primary bottleneck — the model was still actively converging.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3109 because ADVISOR: Closing — lr=4e-4 full-eval only trained 50 epochs (7.2 min/ep with full eval vs 0.7 min/ep truncated). val=12.03% not comparable. Full-eval mode fundamentally incompatible with epoch budget. Guts reassigned..
